src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Split.cs

```diff
@@ -0,0 +1,152 @@
+        /// using the provided separator span.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separator">The separator span to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, ReadOnlySpan<T> separator)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separator, treatAsSingleSeparator: true);
+
+        /// <summary>
+        /// Returns a type that allows for enumeration of each element within a split span
+        /// using any of the provided elements.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separators">The separators to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, [UnscopedRef] params ReadOnlySpan<T> separators)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separators, treatAsSingleSeparator: false);
+
+        /// <summary>
+        /// Returns a type that allows for enumeration of each element within a split span
+        /// using the provided <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separators">The <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/> to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, SearchValues<T> separators)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separators);
+
+        /// <summary>
+        /// <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/> allows for enumeration of each element within a <see cref="System.ReadOnlySpan{T}"/>
+        /// that has been split using a provided separator.
+        /// </summary>
+        public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
+        {
+            private enum SplitMode
+            {
+                None = 0,
+                SingleToken,
+                Sequence,
+                Any,
+                SearchValues
+            }
+
+            private readonly ReadOnlySpan<T> _span;
+
+            private readonly T _separator = default!;
+            private readonly ReadOnlySpan<T> _separatorBuffer;
+            private readonly SearchValues<T> _searchValues = default!;
+
+            private readonly int _separatorLength;
```

Comment:
This shouldn't be necessary. A local in MoveNext can just be set in the switch based on the case.

Reply:
I guess its not strictly necessary, but it would incur a cost for each call of `MoveNext` (that will always evaluate to the same result).

Reply:
>  but it would incur a cost for each call of MoveNext

There shouldn't be an extra branch, as that'd already be covered by the switch, and this would either be storing a constant into a local or storing a span's Length into a local. That versus loading a field. Might be worth measuring, but I wouldn't be surprised if it were actually cheaper.

Reply:
There would be an extra cost for the `SplitMode.Sequence` scenario, as it will have to determine whether the `_separatorBuffer` is of length `0`, in which case the separatorLength would be `1`. See [here](https://gist.github.com/bbartels/167d2e6ecb8979a7aa4db360263bdb2f#file-spansplit_movenext_separatorlengthcalc-cs-L13). It would break the following testcase: https://github.com/dotnet/runtime/pull/104534/files#diff-27da3246e86c068a251dfa98f5c889b71a6d58d6022040c36c99442fa881d529R121

Reply:
> There would be an extra cost for the SplitMode.Sequence scenario, as it will have to determine whether the _separatorBuffer is of length 0, in which case the separatorLength would be 1

You could just have an extra enum value to represent that.

Reply:
Hadn't considered it being a discrete SplitMode, amended!